### PR TITLE
Uncertainty Quantification with PET-MAD

### PR DESCRIPTION
**Pull request by johannes-spies**
Three examples of using PET-MAD's uncertainty quantification.

### CODE REVIEW

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+plt.grid()
+plt.gca().set(
+    aspect="equal",
+    title="Parity of Uncertainties",
+    ylabel="Errors",
+    xlabel="Uncertainties",
+)
+plt.loglog()
+plt.plot([min_val, max_val], [min_val, max_val], ls="--", c="k")
+plt.scatter(predicted_uncertainties, groundtruth_uncertainties)
```
Comment (SanggyuChong) [inline examples/pet-mad-uq/pet-mad-uq.py]: The plot could be improved! please "zoom in" to the range of predicted uncertainties, and it would be nice to also provide the isolines for the different factors of sigma (I will share the code to do this with you)

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+N = len(supercell)  # store the number of atoms
+
+# Attach the loaded model
+calculator = MetatomicCalculator(model)
+supercell.calc = calculator
+
+# %%
+# Now we can define the requested model outputs. Note that we explicitly request the
+# last-layer features, which are required to estimated the uncertainty, and the
+# ensemble energies, to have another metric for estimating errors.
```
Comment (SanggyuChong) [inline examples/pet-mad-uq/pet-mad-uq.py]: "Note that we explicitly request the last-layer features, which are required to estimated the uncertainty, and the ensemble energies, to have another metric for estimating errors." --> "Along with the energy_uncertainty and energy_ensemble, we also explicitly request for the last layer features, which are needed in the analytical computation of the uncertainties of the vacancy formation energies that we attempt further down"

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 3]
```diff
@@ -0,0 +1,518 @@
...
+
+output_options = {
+    # request the uncertainty in the atomic energy predictions
+    "energy": ModelOutput(),  # needed to request the uncertainties
+    "energy_uncertainty": ModelOutput(),
+    "mtt::aux::energy_last_layer_features": ModelOutput(),
+    "energy_ensemble": ModelOutput(),  # needed to request energy ensemble predictions
```
Comment (SanggyuChong) [inline examples/pet-mad-uq/pet-mad-uq.py]: Also, this is the first instance where the ensemble is discussed. Would be nice to inform the reader how many members there are!

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+vacancy_formation_energy_ensemble = torch.mean(
+    vacancy[1]["energy_ensemble"] - (N - 1) / N * bulk[1]["energy_ensemble"]
+)
+vacancy_formation_energy_ensemble_var = torch.var(
+    vacancy[1]["energy_ensemble"] - (N - 1) / N * bulk[1]["energy_ensemble"]
+)
+
+# %%
+# The uncertainty in the vacancy formation energy cannot be computed in a similar,
+# straightforward way similar to the vacancy formation energy. As mentioned previously,
```
Comment (SanggyuChong) [inline examples/pet-mad-uq/pet-mad-uq.py]: I think the explanation here can be a lot more clear. Will discuss with you in person to suggest edits :)

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+    for field in ["last_layer_features", "energy_ensemble"]:
+        if field in vs:
+            del vs[field]
+    for field in vs:
+        if isinstance(vs[field], torch.Tensor):
+            vs[field] = vs[field].item()
+
+# %%
+# The following table summarizes the values computed using different methods.
+pd.DataFrame(values, index=index)
```
Comment (SanggyuChong) [inline examples/pet-mad-uq/pet-mad-uq.py]: Wouldn't it be nicer to show this as bar graphs with error bars?

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 1 above)
Comment (SanggyuChong) [reply examples/pet-mad-uq/pet-mad-uq.py]: Also, people may struggle with the interpretation of this plot. I also suggest a reference to one of the appendices in Bigi et al. that helps with the interpretation

**Review by SanggyuChong — CHANGES_REQUESTED** [review]
Looks very nice, great work Johannes! Left some minor comments for edits, and I'll come to you to explain them further

**Comment by SanggyuChong** [comment]
Upon further discussion, @johannes-spies will also provide a brief description of how the LLPR-wrapped model can be obtained using the metatrain helper functions

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 5 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: Yes, but that also does not super enhance the readability. For now, I just rounded the numbers to make the table less crowded, as we discussed. I've addressed this in 828d203.

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 4 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: Thanks, I tried again in 668aa5f.

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 3 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: Good catch! I've changed that in 96e243e.

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 1 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: Fixed in a8e6655 and abe3a8a.

**Review by SanggyuChong — APPROVED** [review]
I have suggested a few minor text edits offline, but otherwise this looks good to me :confetti_ball:

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+The LLPR uncertainties are introduced in `Bigi et al., 2024.
+<https://arxiv.org/abs/2403.02251>`_ For more
+information on dataset calibration and error propagation, see
+`Imabalzano et al., 2021. <https://arxiv.org/abs/2011.08828>`_
+
+Models compatible with `metatomic <https://metatensor.github.io/metatomic/>`_ can be
+equipped with uncertainty quantification capabilities through the
+`LLPRUncertaintyModel` wrapper included with
+`metatrain <https://metatensor.github.io/metatrain/>`_. For running this recipe, you can
+use a prebuilt model (the example itself downloads a model from Hugging Face).
```
Comment (frostedoyster) [inline examples/pet-mad-uq/pet-mad-uq.py]: I think this part (where we show how to add UQ to a metatomic model) should go into a big section with its own header (i.e., the title of the section). In particular, I think it's important to make sure that the users understand that they don't have to run this part. A good title for this header could be "Adding UQ to a model (not required)".

Comment (frostedoyster) [reply examples/pet-mad-uq/pet-mad-uq.py]: This is important because our API to add UQ to a model is also about to change

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+from metatrain.utils.data.system_to_ase import system_to_ase
+
+
+# %%
+# Model Preparation
+# -----------------
+# All examples require a PET-MAD model with ensemble and LLPR prediction. The
+# following
+# code loads a pre-built model using the ASE-compatible calculator wrapper. Using the
+# calculator instead of calling the model directly conveniently hides computing
```
Comment (frostedoyster) [inline examples/pet-mad-uq/pet-mad-uq.py]: Here, I don't think it's clear enough that the model is pre-trained and that the user doesn't have to train anything related to the model or the LLPR. I think it would be good to change the title to "Model loading" or "Loading the model" and to add the pre-trained word (e.g. instead of "pre-built")

**Review by frostedoyster — COMMENTED** [review]
Very good PR and examples, thanks for all the work @johannes-spies
My comments are mostly about not scaring the users off with the code snippet at the beginning, which is messy and probably unstable (in the sense that it will change)

File: examples/pet-mad-uq/pet-mad-uq.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+# neighbor lists in the calculator. By default, PET-MAD does not (yet) come equipped
```
Comment (frostedoyster) [inline examples/pet-mad-uq/pet-mad-uq.py]: By the way, this is not true anymore. We changed the structure to publish a single LLPR-enabled model. There will be no published model without UQ starting from the next version of PET-MAD. Please verify with @abmazitov. We can also wait until that model is available through HuggingFace

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 6 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: True, thanks. I added an "Optional: Adding UQ to a Model" to the respective section in 89b3bd0.

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 7 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: Done in 89b3bd0.

File: examples/pet-mad-uq/pet-mad-uq.py (hunk 8 above)
Comment (johannes-spies) [reply examples/pet-mad-uq/pet-mad-uq.py]: I removed the last sentence of the section in 89b3bd0 for now. But you're right, this is something that needs to be changed when the UQ-enabled model becomes available on Hugging Face.

**Comment by frostedoyster** [comment]
The recipe is good in my opinion. I will create an issue to update it to the newest PET-MAD as soon as it's on HF and a new version of metatrain is released
